## You're here! 
Welcome to your first competition in the [ITI's AI Pro training program](https://ai.iti.gov.eg/epita/ai-engineer/)! We hope you enjoy and learn as much as we did prepairing this competition.


## Introduction

In the competition, it's required to predict the `Severity` of a car crash given info about the crash, e.g., location.

This is the getting started notebook. Things are kept simple so that it's easier to understand the steps and modify it.

Feel free to `Fork` this notebook and share it with your modifications **OR** use it to create your submissions.

### Prerequisites
You should know how to use python and a little bit of Machine Learning. You can apply the techniques you learned in the training program and submit the new solutions! 

### Checklist
You can participate in this competition the way you perefer. However, I recommend following these steps if this is your first time joining a competition on Kaggle.

* Fork this notebook and run the cells in order.
* Submit this solution.
* Make changes to the data processing step as you see fit.
* Submit the new solutions.

*You can submit up to 5 submissions per day. You can select only one of the submission you make to be considered in the final ranking.*


Don't hesitate to leave a comment or contact me if you have any question!

## Import the libraries

We'll use `pandas` to load and manipulate the data. Other libraries will be imported in the relevant sections.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt 
import os
import numpy as np

## Exploratory Data Analysis
In this step, one should load the data and analyze it. However, I'll load the data and do minimal analysis. You are encouraged to do thorough analysis!

Let's load the data using `pandas` and have a look at the generated `DataFrame`.

In [2]:
from functools import partialmethod
pd.DataFrame.head = partialmethod(pd.DataFrame.head, n=1)



In [8]:
dataset_path = 'kaggle/input/car-crashes-severity-prediction/'

df = pd.read_csv(os.path.join(dataset_path, 'train.csv'))

print("The shape of the dataset is {}.\n\n".format(df.shape))

df.head()

The shape of the dataset is (6407, 16).




ID       Lat        Lng   Bump  Distance(mi)  Crossing  Give_Way  Junction  \
0   0  37.76215 -122.40566  False         0.044     False     False     False   

   No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0    False    False       False  False     True    R         2   

             timestamp  
0  2016-03-25 15:13:02

In [9]:
df['timestamp'] = pd.to_datetime(df['timestamp'])
df['date'] = df['timestamp'].dt.date
df['hour'] = df['timestamp'].dt.strftime('%H')
df['hours_Min'] = df['timestamp'].dt.strftime('%H:%M')

In [10]:
df['Severity'].value_counts()

2    4346
3    1855
1     129
4      77
Name: Severity, dtype: int64

In [11]:
df.head()

ID       Lat        Lng   Bump  Distance(mi)  Crossing  Give_Way  Junction  \
0   0  37.76215 -122.40566  False         0.044     False     False     False   

   No_Exit  Railway  Roundabout   Stop  Amenity Side  Severity  \
0    False    False       False  False     True    R         2   

            timestamp        date hour hours_Min  
0 2016-03-25 15:13:02  2016-03-25   15     15:13

In [ ]:
# this cell is used to convert T/F features into 0/1 features

df['Bump'] = df['Bump'].astype(int)
df['Crossing'] = df['Crossing'].astype(int)
df['Give_Way'] = df['Give_Way'].astype(int)
df['Junction'] = df['Junction'].astype(int)
df['No_Exit'] = df['No_Exit'].astype(int)
df['Railway'] = df['Railway'].astype(int)
df['Roundabout'] = df['Roundabout'].astype(int)
df['Stop'] = df['Stop'].astype(int)
df['Amenity'] = df['Amenity'].astype(int)
df['hour'] = df['hour'].astype(int)
# this cell is used to convert the Side column from L/R data into 0/1 data

df['Side'] = df['Side'].map(dict(L=0, R=1))

In [ ]:
df.head()

In [ ]:
fig, axes = plt.subplots(nrows=2, ncols=2, figsize=(20,10))
for (serv, ax) in zip(np.arange(1,5), axes.flatten()):
    df[[ 'hour' , "Severity" ]].loc[df['Severity'] == serv].groupby(['hour']).agg(['count']).plot(kind = "bar" ,ax =ax)    
    ax.legend(f"count")
    ax.set_title(f"Severity {str(serv)} Count vs Hours of the day")


In [ ]:
df[[ 'hour' , "Severity" ]].loc[df['Severity'] == 1].groupby(['hour']).agg(['count']).plot(kind = "bar")
df[[ 'hour' , "Severity" ]].loc[df['Severity'] == 2].groupby(['hour']).agg(['count']).plot(kind = "bar")
df[[ 'hour' , "Severity" ]].loc[df['Severity'] == 3].groupby(['hour']).agg(['count']).plot(kind = "bar")
df[[ 'hour' , "Severity" ]].loc[df['Severity'] == 4].groupby(['hour']).agg(['count']).plot(kind = "bar")

In [ ]:
plt.scatter(x = df['hour'] , y = df['Severity'] , alpha = .2)

### reading holidays xml

In [ ]:
holidays_df = pd.read_xml(os.path.join(dataset_path, 'holidays.xml'))

print("The shape of the dataset is {}.\n\n".format(holidays_df.shape))

holidays_df

### reading weather csv

In [ ]:
weather_df = pd.read_csv(os.path.join(dataset_path, 'weather-sfcsv.csv'))

print("The shape of the dataset is {}.\n\n".format(weather_df.shape))

weather_df.head()

In [ ]:
weather_df['timestamp'] = pd.to_datetime(weather_df[['Year', 'Month','Day','Hour']])
weather_df['date'] = weather_df['timestamp'].dt.date
weather_df.head()

In [ ]:
weather_df.head()

We've got 6407 examples in the dataset with 14 featues, 1 ID, and the `Severity` of the crash.

By looking at the features and a sample from the data, the features look of numerical and catogerical types. What about some descriptive statistics?

### Merging the Data into one DataFrame

In [ ]:
df.head()

In [ ]:
df.info()

In [ ]:
weather_df.info()

In [ ]:
DfwithWeatherData = pd.merge(df,weather_df,how = "left" , left_on=['date','hour'],right_on=['date','Hour'] )
DfwithWeatherData

In [ ]:
holidays_df.info()

In [ ]:
merged = pd.merge(DfwithWeatherData,holidays_df ,how = "left" , left_on=['date'],right_on=['date'] )
merged
for i in merged['date']:
    if i in holidays_df['date']:
         merged.loc['date' == i, 'description'] = holidays_df.loc['date' == i, 'description']

In [ ]:
merged.columns

In [ ]:
merged.info()

In [ ]:
holidays_df

In [ ]:
merged.info()

In [ ]:
merged['description'].unique()

In [ ]:
merged.loc[merged['date'] == '2016-03-25']

In [ ]:
merged[['date', 'Weather_Condition','description']].loc[merged['description']!=np.nan]

In [ ]:
merged.to_csv("fi2.csv")

### Data Cleaning

In [ ]:
df.drop(columns='ID').describe()

The output shows desciptive statistics for the numerical features, `Lat`, `Lng`, `Distance(mi)`, and `Severity`. I'll use the numerical features to demonstrate how to train the model and make submissions. **However you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.**

In [ ]:
DfwithWeatherData.drop(columns=['ID', 'Severity'])

In [ ]:
dnew = DfwithWeatherData.dropna()
min_dis = dnew['Distance(mi)'].min()
max_dis = dnew['Distance(mi)'].max()
dnew['Distance(mi)_norm'] = (dnew['Distance(mi)'] - min_dis)/(max_dis - min_dis)
min_sp = dnew['Wind_Speed(mph)'].min()
max_sp = dnew['Wind_Speed(mph)'].max()
dnew['Wind_Speed(mph)_norm'] = (dnew['Wind_Speed(mph)'] - min_sp)/(max_sp - min_sp)
min_vis = dnew['Visibility(mi)'].min()
max_vis = dnew['Visibility(mi)'].max()
dnew['Visibility(mi)_norm'] = (dnew['Visibility(mi)'] - min_vis)/(max_vis - min_vis)
min_chill = dnew['Wind_Chill(F)'].min()
max_chill = dnew['Wind_Chill(F)'].max()
dnew['Wind_Chill(F)_norm'] = (dnew['Wind_Chill(F)'] - min_chill)/(max_chill - min_chill)
dnew['Side'].mean()

## Data Splitting

Now it's time to split the dataset for the training step. Typically the dataset is split into 3 subsets, namely, the training, validation and test sets. In our case, the test set is already predefined. So we'll split the "training" set into training and validation sets with 0.8:0.2 ratio. 

*Note: a good way to generate reproducible results is to set the seed to the algorithms that depends on randomization. This is done with the argument `random_state` in the following command* 

In [ ]:
from sklearn.model_selection import train_test_split

train_df, val_df = train_test_split(dnew, test_size=0.2, random_state=42) # Try adding `stratify` here

X_train = train_df.drop(columns=['ID', 'Severity'])
y_train = train_df['Severity']

X_val = val_df.drop(columns=['ID', 'Severity'])
y_val = val_df['Severity']


As pointed out eariler, I'll use the numerical features to train the classifier. **However, you shouldn't use the numerical features only to make the final submission if you want to make it to the top of the leaderboard.** 

In [ ]:
X_train

In [ ]:
X_train.shape

In [ ]:
# This cell is used to select the numerical features. IT SHOULD BE REMOVED AS YOU DO YOUR WORK.
X_train = X_train[['Lat', 'Lng', 'Distance(mi)_norm','Junction','Stop','Amenity','Side','Wind_Speed(mph)_norm','Visibility(mi)_norm']]
X_val = X_val[['Lat', 'Lng', 'Distance(mi)_norm','Junction','Stop','Amenity','Side','Wind_Speed(mph)_norm','Visibility(mi)_norm']]

In [ ]:
X_train

## Model Training

Let's train a model with the data! We'll train a Random Forest Classifier to demonstrate the process of making submissions. 

In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Create an instance of the classifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)

# Train the classifier
classifier = classifier.fit(X_train, y_train)

Now let's test our classifier on the validation dataset and see the accuracy.

In [ ]:
print("The accuracy of the classifier on the validation set is ", (classifier.score(X_val, y_val)))

Well. That's a good start, right? A classifier that predicts all examples' `Severity` as 2 will get around 0.63. You should get better score as you add more features and do better data preprocessing.

## Submission File Generation

We have built a model and we'd like to submit our predictions on the test set! In order to do that, we'll load the test set, predict the class and save the submission file. 

First, we'll load the data.

In [ ]:
test_df = pd.read_csv(os.path.join(dataset_path, 'test.csv'))
test_df.head()

Note that the test set has the same features and doesn't have the `Severity` column.
At this stage one must **NOT** forget to apply the same processing done on the training set on the features of the test set.

Now we'll add `Severity` column to the test `DataFrame` and add the values of the predicted class to it.

**I'll select the numerical features here as I did in the training set. DO NOT forget to change this step as you change the preprocessing of the training data.**

In [ ]:
X_test = test_df.drop(columns=['ID'])

# You should update/remove the next line once you change the features used for training
X_test = X_test[['Lat', 'Lng', 'Distance(mi)']]

y_test_predicted = classifier.predict(X_test)

test_df['Severity'] = y_test_predicted

test_df.head()

Now we're ready to generate the submission file. The submission file needs the columns `ID` and `Severity` only.

In [ ]:
test_df[['ID', 'Severity']].to_csv('/kaggle/working/submission.csv', index=False)

The remaining steps is to submit the generated file and are as follows. 

1. Press `Save Version` on the upper right corner of this notebook.
2. Write a `Version Name` of your choice and choose `Save & Run All (Commit)` then click `Save`.
3. Wait for the saved notebook to finish running the go to the saved notebook.
4. Scroll down until you see the output files then select the `submission.csv` file and click `Submit`.

Now your submission will be evaluated and your score will be updated on the leaderboard! CONGRATULATIONS!!

## Conclusion

In this notebook, we have demonstrated the essential steps that one should do in order to get "slightly" familiar with the data and the submission process. We chose not to go into details in each step to keep the welcoming notebook simple and make a room for improvement.

You're encourged to `Fork` the notebook, edit it, add your insights and use it to create your submission.